scylla/src/transport/node.rs

```diff
@@ -139,13 +139,32 @@ impl Node {
                 shard_info,
                 shard_conns,
             } => {
-                let shard = rand::thread_rng().gen_range(0..shard_info.nr_shards as u32);
-                // It's guaranteed by NodeWorker that shard_conns.len() == shard_info.nr_shards
-                shard_conns[shard as usize].get_connection().await
+                let shard: usize = rand::thread_rng().gen_range(0..shard_info.nr_shards).into();
+                Self::connection_for_shard(shard, shard_conns).await
             }
         }
     }
 
+    // Tries to get a connection to given shard, if it's broken returns any working connection
+    async fn connection_for_shard(
+        shard: usize,
+        shard_conns: &[ConnectionKeeper],
+    ) -> Result<Arc<Connection>, QueryError> {
```

Comment:
:+1: fixed. I did it in a semi optimal way by collecting all shards to a `Vec` and shuffling them randomly.

There is a crate that could provide a constant memory iteration through a random permutation using some clever mathematical tricks: [`permutation_iterator`](https://crates.io/crates/permutation_iterator), but I'm not sure if we want to depend on it:
* It has only 590 downloads all time
* It has a single author, can we trust them?
* It has an Apache license, we have double MIT and APACHE, is this a problem? I really need to read up on licenses.

Reply:
I think that we don't need to include that dependency. The happy path where we do have a connection for the given shard does not allocate a `Vec`. The "sad" path shouldn't be too bad - the number of shards is usually small (should usually be below 100), so the allocation shouldn't be too big.

As for trying shards in random order, I think permuting the whole vector up front is not necessary as you probably won't need to try all connections most of the time. You can save some calls to the random generator by drawing shards on demand in a loop:

```rust
while !shards_to_try.is_empty() {
    let idx = rand::thread_rng().gen_range(0..shards_to_try.len());
    let shard = shards_to_try.swap_remove(idx);

    match shard_conns[shard].get_connection().await {
        Ok(conn) => return Ok(conn),
        Err(e) => last_error = e,
    }
}
```

I also think that you could get away with using `u16` to represent shard IDs. This would make the `Vec` 4x smaller (on 64bit architectures, at least).